libsolidity/formal/Predicate.h

```diff
@@ -0,0 +1,124 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/formal/SymbolicVariables.h>
+#include <libsolidity/formal/SymbolicVariables.h>
+
+#include <libsmtutil/Sorts.h>
+
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace solidity::frontend
+{
+
+/**
+ * Represents a predicate used by the CHC engine.
+ */
+class Predicate
+{
+public:
+	static Predicate const* create(
+		smtutil::SortPointer _sort,
+		std::string _name,
+		smt::EncodingContext& _context,
+		ASTNode const* _node = nullptr
+	);
+
+	/// Predicate should be movable.
+	Predicate(Predicate&&) = default;
```

Comment:
Wouldn't compile without this

Reply:
Haha, yeah, just commented about this - I think this should be deleted as well and you should change the map emplace to an in-place emplace, then it should still compile without being even movable.

Reply:
Ah, hm... you need to construct it first to get the name under which to store it in the map the way things are built right now? Hm... that's not so nice... let me see...

Reply:
Yes exactly

Reply:
I just put up a lengthier suggestion - since the constructor is private you can just as well construct the ``SymbolicFunctionVariable`` in ``create`` instead and just move it down the constructor... I think that should work.

Reply:
An alternative would be to make the static map actually a ``set`` that has a transparent comparison operator that compares according to the name (and being transparent allows lookups by name) :-D... but that might be extreme overkill :-) (wanting the move constructor deleted may already qualify as slight overkill ;-)).